scrapy/core/engine.py

```diff
@@ -102,188 +109,175 @@ def close(self):
                                            exc_info=failure_to_exc_info(f),
                                            extra={'spider': spider}))
-        d.addBoth(lambda _: slot.remove_request(request))
+        d.addBoth(lambda _: self.slot.remove_request(request))
         d.addErrback(lambda f: logger.info('Error while removing request from slot',
                                            exc_info=failure_to_exc_info(f),
                                            extra={'spider': spider}))
-        d.addBoth(lambda _: slot.nextcall.schedule())
+        d.addBoth(lambda _: self.slot.nextcall.schedule())
         d.addErrback(lambda f: logger.info('Error while scheduling new request',
                                            exc_info=failure_to_exc_info(f),
                                            extra={'spider': spider}))
         return d
 
-    def _handle_downloader_output(self, response, request, spider):
-        if not isinstance(response, (Request, Response, Failure)):
-            raise TypeError(
-                "Incorrect type: expected Request, Response or Failure, got "
-                f"{type(response)}: {response!r}"
-            )
+    def _handle_downloader_output(
+        self, result: Union[Request, Response, Failure], request: Request, spider: Spider
+    ) -> Optional[Deferred]:
+        if not isinstance(result, (Request, Response, Failure)):
+            raise TypeError(f"Incorrect type: expected Request, Response or Failure, got {type(result)}: {result!r}")
         # downloader middleware can return requests (for example, redirects)
-        if isinstance(response, Request):
-            self.crawl(response, spider)
-            return
-        # response is a Response or Failure
-        d = self.scraper.enqueue_scrape(response, request, spider)
+        if isinstance(result, Request):
+            self.crawl(result, spider)
+            return None
+        d = self.scraper.enqueue_scrape(result, request, spider)
         d.addErrback(lambda f: logger.error('Error while enqueuing downloader output',
                                             exc_info=failure_to_exc_info(f),
                                             extra={'spider': spider}))
         return d
 
-    def spider_is_idle(self, spider):
+    def spider_is_idle(self, spider: Spider) -> bool:
         if not self.scraper.slot.is_idle():
-            # scraper is not idle
             return False
-
-        if self.downloader.active:
-            # downloader has pending requests
+        if self.downloader.active:  # downloader has pending requests
             return False
-
-        if self.slot.start_requests is not None:
-            # not all start requests are handled
-            return False
-
-        if self.slot.scheduler.has_pending_requests():
-            # scheduler has pending requests
-            return False
-
+        if self.slot is not None:
```

Comment:
I think it makes sense to check here, given `spider_is_idle` is a public method. But since all current code expects it not to be `None`, I wonder if we should instead raise an exception, e.g. a `RuntimeError` complaining about `spider_is_idle` being called before `open_spider`.

Same for `crawl` and `schedule` below.